File: arrays-and-hashing/java-solutions/GroupAnagrams.java
import java.util.*;

public class GroupAnagrams {
     public static void main(String[] args) {
          System.out.println(groupedAnagrams(new String[]{"eat","tea","tan","ate","nat","bat"}));
          return;
     }

     public static ArrayList<ArrayList<String>> groupedAnagrams(String[] words) {
          
          HashMap<String, ArrayList<String>> anagrams = new HashMap<String, ArrayList<String>>();

          for (String word : Arrays.asList(words)) {
               char[] key = word.toCharArray();
               Arrays.sort(key);
               

               if (anagrams.containsKey(new String(key))) {
                    anagrams.get(new String(key)).add(word);
               } else {
                    anagrams.put(new String(key), new ArrayList<String>(Arrays.asList(word)));
               }
          }

          ArrayList<ArrayList<String>> grouped = new ArrayList<ArrayList<String>>();

          for (ArrayList<String> group : anagrams.values()) {
               grouped.add(group);
          }

          return grouped;
     }
}

File: arrays-and-hashing/group-anagrams.py
def group_anagrams(words):
     """
     Groups a list of words into anagrams.

     Parameters:
     words (list): A list of strings representing words.

     Returns:
     list: A list of lists, where each inner list contains words that are anagrams of each other.
     """
     hashMap = {}
     result = []

     for i in words:
          sortedStr = "".join(sorted(i))
          if (not sortedStr in hashMap):
               hashMap[sortedStr] = [i]
          else:
               hashMap[sortedStr].append(i)

     # append all remaining values
     if hashMap.values():
          for j in hashMap.values():
               result.append(j)
          
     return result


print(group_anagrams(["eat","tea","tan","ate","nat","bat"]))
